Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/calibration/ErrorHandler.java

```diff
@@ -0,0 +1,519 @@
+    protected RangeVector[] pastForecasts;
+
+    RangeVector errorDistribution;
+    DiVector errorRMSE;
+    float[] errorMean;
+    Deviation[] intervalPrecision;
+    Deviation[] rmseHighDeviations;
+    Deviation[] rmseLowDeviations;
+    float[] lowerLimit;
+    float[] upperLimit;
+    double[] lastInputs;
+    PredictiveRandomCutForest estimator;
+    float[] lastDataDeviations;
+
+    // We keep the multipliers defined for potential
+    // future use.
+
+    RangeVector multipliers;
+    RangeVector adders;
+
+    public ErrorHandler(Builder builder) {
+        checkArgument(builder.forecastHorizon > 0, "has to be positive");
+        checkArgument(builder.errorHorizon >= builder.forecastHorizon,
+                "intervalPrecision horizon should be at least as large as forecast horizon");
+        checkArgument(builder.errorHorizon <= 1024, "reduce error horizon");
+        forecastHorizon = builder.forecastHorizon;
+        errorHorizon = builder.errorHorizon;
+        int inputLength = (builder.dimensions / builder.shingleSize);
+        int length = inputLength * forecastHorizon;
+        percentile = builder.percentile;
+        pastForecasts = new RangeVector[forecastHorizon];
+        for (int i = 0; i < forecastHorizon; i++) {
+            pastForecasts[i] = new RangeVector(length);
+        }
+        sequenceIndex = 0;
+        lastInputs = new double[2 * inputLength];
+        rmseHighDeviations = new Deviation[length];
+        rmseLowDeviations = new Deviation[length];
+        intervalPrecision = new Deviation[length];
+        for (int i = 0; i < length; i++) {
+            rmseHighDeviations[i] = new Deviation(1.0 / errorHorizon);
+            rmseLowDeviations[i] = new Deviation(1.0 / errorHorizon);
+            intervalPrecision[i] = new Deviation(1.0 / errorHorizon);
+        }
+        errorMean = new float[length];
+        errorRMSE = new DiVector(length);
+        lastDataDeviations = new float[inputLength];
+        errorDistribution = new RangeVector(length);
+        if (builder.upperLimit.isPresent()) {
+            checkArgument(builder.upperLimit.get().length == inputLength, "incorrect length");
+            upperLimit = Arrays.copyOf(builder.upperLimit.get(), inputLength);
+        } else {
+            upperLimit = new float[inputLength];
+            Arrays.fill(upperLimit, Float.MAX_VALUE);
+        }
+        if (builder.lowerLimit.isPresent()) {
+            checkArgument(builder.lowerLimit.get().length == inputLength, "incorrect length");
+            for (int y = 0; y < inputLength; y++) {
+                checkArgument(builder.lowerLimit.get()[y] <= upperLimit[y], "incorrect limits");
+            }
+            lowerLimit = Arrays.copyOf(builder.lowerLimit.get(), inputLength);
+        } else {
+            lowerLimit = new float[inputLength];
+            Arrays.fill(lowerLimit, -Float.MAX_VALUE);
+        }
+        // uses lastInputs as a markov input, the +1 corresponds to lookahead and then
```

Comment:
You did +2 in the code.

Reply:
Yes - it's lookahead and forecastHorizon - lookahead, as we discussed in previous comment. You're correct the comment needs to be changed. Let's come back to this after we have gone though the other code? Otherwise it seems I am generating too many commits :)

Reply:
Done